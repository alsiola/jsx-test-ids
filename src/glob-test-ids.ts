import * as _glob from "glob";
import * as _path from "path";
import * as _fs from "fs-extra";

export const globTestIds = ({ fs = _fs, glob = _glob, path = _path }) => ({
    idsLocation,
    output
}: {
    idsLocation: string;
    output: string;
}): void => {
    try {
        const allIds = glob
            .sync(path.join(process.cwd(), idsLocation))
            .map(filename => ({
                filename,
                file: fs.readFileSync(filename, "utf8")
            }))
            .map(({ filename, file }) => ({
                filename,
                file: JSON.parse(file) as string[]
            }))
            .reduce((ids, { filename, file }) => {
                file.forEach(testId => {
                    if (testId in ids) {
                        throw new Error(
                            `Duplicate test id ${testId} in ${filename}`
                        );
                    }
                    ids[testId] = testId;
                });
                return ids;
            }, {} as Record<string, string>);

        fs.mkdirpSync(path.dirname(output));

        const outFile = path.join(process.cwd(), output);

        fs.writeFileSync(outFile, JSON.stringify(allIds, null, 2));
        console.log(
            `${Object.keys(allIds).length} test ids written to ${outFile}`
        );
        process.exit(0);
    } catch (e) {
        console.error(e);
        process.exit(1);
    }
};
