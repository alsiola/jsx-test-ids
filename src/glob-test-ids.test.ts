import { globTestIds } from "./glob-test-ids";

const setupMocks = ({ testIds }: { testIds: any }) => {
    const fs = {
        readFileSync: jest.fn().mockReturnValue(JSON.stringify(testIds)),
        mkdirpSync: jest.fn(),
        writeFileSync: jest.fn()
    };
    const path = {
        join: jest.fn().mockReturnValue("outfile"),
        dirname: jest.fn().mockReturnValue("dirname")
    };
    const glob = {
        sync: jest.fn().mockReturnValue(["filename"])
    };

    const fn = globTestIds({ fs, glob, path } as any);

    return {
        fs,
        path,
        glob,
        fn
    };
};

describe("globTestIds", () => {
    it("reads all test ids and writes to json file", () => {
        const { fs, path, glob, fn } = setupMocks({ testIds: ["test-id"] });

        fn({
            idsLocation: "./test",
            output: "./master.json",
            cwd: "/test/dir/"
        });

        expect(fs.readFileSync.mock.calls).toMatchSnapshot("fs.readFileSync");
        expect(fs.mkdirpSync.mock.calls).toMatchSnapshot("fs.mkdirpSync");
        expect(fs.writeFileSync.mock.calls).toMatchSnapshot("fs.writeFileSync");
        expect(path.join.mock.calls).toMatchSnapshot("path.join");
        expect(path.dirname.mock.calls).toMatchSnapshot("path.dirname");
        expect(glob.sync.mock.calls).toMatchSnapshot("glob.sync");
    });

    it("throws if there are duplicate test ids", () => {
        const { fn } = setupMocks({
            testIds: ["test-id", "test-id"]
        });

        expect(() =>
            fn({
                idsLocation: "./test",
                output: "./master.json",
                cwd: "/test/dir/"
            })
        ).toThrowErrorMatchingSnapshot();
    });
});
