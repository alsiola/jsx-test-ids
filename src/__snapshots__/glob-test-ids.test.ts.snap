// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`globTestIds reads all test ids and writes to json file: fs.mkdirpSync 1`] = `
Array [
  Array [
    "dirname",
  ],
]
`;

exports[`globTestIds reads all test ids and writes to json file: fs.readFileSync 1`] = `
Array [
  Array [
    "filename",
    "utf8",
  ],
]
`;

exports[`globTestIds reads all test ids and writes to json file: fs.writeFileSync 1`] = `
Array [
  Array [
    "outfile",
    "{
  \\"test-id\\": \\"test-id\\"
}",
  ],
]
`;

exports[`globTestIds reads all test ids and writes to json file: glob.sync 1`] = `
Array [
  Array [
    "outfile",
  ],
]
`;

exports[`globTestIds reads all test ids and writes to json file: path.dirname 1`] = `
Array [
  Array [
    "./master.json",
  ],
]
`;

exports[`globTestIds reads all test ids and writes to json file: path.join 1`] = `
Array [
  Array [
    "/test/dir/",
    "./test",
  ],
  Array [
    "/test/dir/",
    "./master.json",
  ],
]
`;

exports[`globTestIds throws if there are duplicate test ids 1`] = `"Duplicate test id test-id in filename"`;
